### PR TITLE
[schema processor] Part 3 - Modifiers and Revisions #12147

### PR DESCRIPTION
**Pull request by MovieStoreGuy**
This is a new PR of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12147

No existing changes have happened.

cc: @tigrannajaryan

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
@MovieStoreGuy I reviewed https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12147 earlier. Does this need another round of review?

**Comment by MovieStoreGuy** [comment]
> @MovieStoreGuy I reviewed #12147 earlier. Does this need another round of review?

This is the same PR but nether of us could reopen the stale PR.

**Comment by MovieStoreGuy** [comment]
I believe I had addressed all the comments left by @tigrannajaryan in the last reviewed pinned.

I know in my delay that schema parser as adopted v1.1, however, I would like to get this out first that works with v1.0 so the semantic convetions can be defined as stable and v1.1 can be added in once this component is added to the release.

File: processor/schemaprocessor/internal/translation/modifier_noop.go
Comment (bogdandrutu) [inline processor/schemaprocessor/internal/translation/modifier_noop.go:24]: nopModifier?

File: processor/schemaprocessor/internal/translation/modifier.go
Comment (bogdandrutu) [inline processor/schemaprocessor/internal/translation/modifier.go:28]: Hard to understand the diff between update and revert. After reading only this interface comments I don't know when to use one or the other, this is a bad sign in my opinion.

File: processor/schemaprocessor/internal/translation/modifier.go
Comment (bogdandrutu) [inline processor/schemaprocessor/internal/translation/modifier.go:49]: Is this a "rename" action?

File: processor/schemaprocessor/internal/translation/modifier.go
Comment (bogdandrutu) [inline processor/schemaprocessor/internal/translation/modifier.go:78]: Duplicate code from 78-97 with 104-123

File: processor/schemaprocessor/internal/translation/modifier.go
Comment (bogdandrutu) [inline processor/schemaprocessor/internal/translation/modifier.go:61]: I expect that changes are "applied" in order.

File: processor/schemaprocessor/internal/translation/revision.go
Comment (bogdandrutu) [inline processor/schemaprocessor/internal/translation/revision.go:28]: Is this not just a combination of all the others? Where is this used?

File: processor/schemaprocessor/internal/translation/revision.go
Comment (bogdandrutu) [inline processor/schemaprocessor/internal/translation/revision.go:160]: I think I now understand better the "actions":

So there is a forward change (from v1 -> v2) and a backwards change (v2 ->v1). Not sure how to name the Modifier interface but clearly needs some naming improvements there.

**Review by bogdandrutu — COMMENTED** [review]
I will think more to suggest some better design/names.

File: processor/schemaprocessor/internal/translation/modifier.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/modifier.go:28]: I wonder if it is worth while writing it as:

```golang
type VersionIterator interface {
    NextAttributes(...)
    PrevAttributes(...)
}
```

Would that help make it clearer?

File: processor/schemaprocessor/internal/translation/revision.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/revision.go:28]: So this is a like for like from the `1.0.0` format, https://opentelemetry.io/docs/reference/specification/schemas/file_format_v1.0.0/#all-section

Basically, if the schemaURL matches then the `all` grouping is applied to every signal.

File: processor/schemaprocessor/internal/translation/modifier.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/translation/modifier.go:77]: I am curious if iterating by `attrs` may result in better performance. Loolkup of `attrs.Get()` likely is slower than `mod.attrs[]` lookup, right, especially if there are a lot of attributes? I am curious how performance of this compares:

```go
attrs.Range(func f(k,v) {
  if v, match := mod.attrs[k]; match {
     ...
  }
})
```

We may also need to verify that the attribute we convert to does not already exist (and error if it does?). 

Also I wonder if we need to make the code work correctly if there are renaming loops (e.g. a->b, b->a). The expectation is that the attributes should be swapped, but the current code will not work correctly in that case. Perhaps we should prohibit that in the spec, currently schemas don't say anything about it.

Finally, since iteration order of maps is not defined, the effects of this function are non-deterministic if there are any renaming loops.

**Review by tigrannajaryan — COMMENTED** [review]
Thanks for working on this @MovieStoreGuy !

File: processor/schemaprocessor/internal/translation/revision.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/revision.go:21]: I had used an interface since at the time the zero value wasn't a safe type.

File: processor/schemaprocessor/internal/translation/modifier_race_test.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/modifier_race_test.go:64]: That is right, just trying to check to see if it does what is expected.

**Comment by MovieStoreGuy** [comment]
Still in progress, I just have been unfortunate with trying to find time to resolve this. 

Talking with @mx-psi is keen to help see the implementation finished and take over my current work.

**Comment by mx-psi** [comment]
Will discuss further with @MovieStoreGuy and have a look at this next week to see how to continue.

@MovieStoreGuy if you have the time, would you mind going through the conversations and marking as resolved any that no longer need addressing?

**Comment by tigrannajaryan** [comment]
@MovieStoreGuy sounds good. Let me know if you need any help from me. I think it is important that we complete this. The lack of the transform processor comes up in our discussions on how to better handle semantic convention changes.

File: processor/schemaprocessor/internal/translation/modifier.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/modifier.go:77]: Potentially, I did forget that `pcommon.Map` is not a conventional golang map so there is a cost there that I would need to come back to.

> We may also need to verify that the attribute we convert to does not already exist (and error if it does?).

That is an interesting question / point, I've not addressed the concern that there is a missing attribute that is defined within the schema. Is there an expected behaviour that is defined within the spec?

> Also I wonder if we need to make the code work correctly if there are renaming loops (e.g. a->b, b->a). The expectation is that the attributes should be swapped, but the current code will not work correctly in that case. Perhaps we should prohibit that in the spec, currently schemas don't say anything about it.

Is this an expected behaviour? I believe I have a test in either this or the next PR that looks to address this in a test where a signal revision has `a --> b` and `b --> a`. Since it only performs one modification per scheme revision. However, I think if you list several modifications within one schema then it could potential happen?
I would need to review this again.

> Finally, since iteration order of maps is not defined, the effects of this function are non-deterministic if there are any renaming loops.

Did you have a inclination of what you'd expect in this scenario? 
My initial thought would be to ignore them and have it so an attribute can only be modified once per revision but I am not fixed to the idea.

File: processor/schemaprocessor/internal/translation/modifier.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/modifier.go:78]: This will more than likely change once I rework the internals of schema modifications.

File: processor/schemaprocessor/internal/translation/revision.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/revision.go:103]: Yeah, I need to revisit the type and include a constructor for it.

File: processor/schemaprocessor/internal/translation/revision.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/revision.go:160]: Yeah, I am still scratching my head on what would be a nice name for the Modifier.

Perhaps something like:
- `SignalValueTranslator` 
- `ValueModifer`
- `RevisionUpdater`

**Comment by MovieStoreGuy** [comment]
Sorry, I was actively working on this so please excuse any broken updates.

**Comment by MovieStoreGuy** [comment]
cc @tigrannajaryan , @mx-psi ,

I've updated the PR here and using more bare metal types and taken the nomenclature that would be used for database migrations when looking at applying changes to values. From this I've created the `migrate` package that contains this updated pattern and `translation.Revision` is where I am expecting exposed types that will be applied to `pdata`.

Please let me know you're thoughts on this <3 

I am hoping once this is merged, we shouldn't be too far away from having this being merged in an indevelopment state to gathering feedback early.

File: processor/schemaprocessor/internal/migrate/doc.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/doc.go:23]:
```suggestion
	StateSelectorApply
```

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/attributes.go:41]: Nit: this probably can be:
```suggestion
type AttributeChangeSetSlice []AttributeChangeSet
```

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/attributes.go:38]:
```suggestion
// AttributeChangeSetSlice allows for `AttributeChangeSet`
```

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/attributes.go:45]:
```suggestion
func NewAttributeChangeSet[Key AttributeKey, Value AttributeKey](mappings map[Key]Value) *AttributeChangeSet {
```

Comment (tigrannajaryan) [reply processor/schemaprocessor/internal/migrate/attributes.go:45]: Does this need to be a generic func? It seems like AttributeKey is always going to be a string. Do we ever need to use a different type?

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/attributes.go:75]: Can we add a TODO to decide what we should do if the key already exists? Should we overwrite it (as it does now) or we should we stop and return an error? This is probably more of a TODO for schema spec.

Comment (tigrannajaryan) [reply processor/schemaprocessor/internal/migrate/attributes.go:75]: The `RemoveIf` contract does not say whether it is ok to modify the attrs from the callback. Can we confirm that this is OK and if so make sure `RemoveIf` comments explicitly bless it?

File: processor/schemaprocessor/internal/migrate/attributes_test.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/attributes_test.go:109]: This appears to rely on undocumented behavior of RemoveIf. I think we either need to document it or change our implementation not to rely on undocumented behavior.

File: processor/schemaprocessor/internal/migrate/conditional.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/conditional.go:32]: Same here, does this need to be generic? It seems only string types are ever needed.

File: processor/schemaprocessor/internal/migrate/conditional.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/conditional.go:43]: Does `values` need to be a slice? It seems we will always call this with one value.

File: processor/schemaprocessor/internal/migrate/signal.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/signal.go:36]: Same here: do we have a case where SignalType is not a string?

File: processor/schemaprocessor/internal/migrate/signal.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/migrate/signal.go:27]: Perhaps:
```suggestion
type SignalNameChange struct {
```

Comment (tigrannajaryan) [reply processor/schemaprocessor/internal/migrate/signal.go:27]: And maybe we can also rename `alias.Signal` to `alias.NamedSignal`.

File: processor/schemaprocessor/internal/translation/revision_v1.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/translation/revision_v1.go:57]: Any particular reason to use 10 as capacity?

File: processor/schemaprocessor/internal/translation/testdata/1.6.1
Comment (tigrannajaryan) [inline processor/schemaprocessor/internal/translation/testdata/1.6.1:7]: The file doesn't see to be used anywhere. Same for testdata/1.9.0

File: processor/schemaprocessor/transformer_test.go
Comment (tigrannajaryan) [inline processor/schemaprocessor/transformer_test.go:43]: Is this deleted intentionally? No longer needed?

**Review by tigrannajaryan — COMMENTED** [review]
I mostly like the change, have just some minor comments/questions.

File: processor/schemaprocessor/internal/translation/revision_v1.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/revision_v1.go:57]: It was just a random number I pulled from no where.

File: processor/schemaprocessor/internal/translation/testdata/1.6.1
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/testdata/1.6.1:7]: It was previously being used in the original code, but I am waiting to the next version to generate revision slice as part of next PR to help keep PRs reviewable.

File: processor/schemaprocessor/transformer_test.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/transformer_test.go:43]: I will need to double check, may have accidentally  caught it in git.

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/attributes.go:75]: Let me follow this one up, because it works and doesn't cause a race condition according to the race detector.

File: processor/schemaprocessor/internal/migrate/conditional.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/conditional.go:32]: This one needs to be generic since the schema package defines the [types](go.opentelemetry.io/otel/schema/v1.0/types) package that forces those values to be new types instead of alias for strings.

The generic receiver here means we can take the `can convert to string` values and store them internally as strings with one cast in the constructer. 
Save me having to write optional handlers for each type defined by the `types` package.

File: processor/schemaprocessor/internal/migrate/conditional.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/conditional.go:43]: That is something I needed to follow up on, one thing I was sure of is in the span events section, you can supply a set for applies to `span event names` and `span names`. I wasn't sure if they can be combined to restrict to specific values or if they are treated differently.

File: processor/schemaprocessor/internal/migrate/signal.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/signal.go:36]: Yes, it is defined here `https://pkg.go.dev/go.opentelemetry.io/otel/schema@v0.0.4/v1.0/ast#MetricsChange` as `map[types.MetricName]types.MetricName` which are new types being defined instead of type alias being used.

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/attributes.go:45]: I've kept it consistent between the other methods since you can't have type constraints without using it as a generic type.

File: processor/schemaprocessor/transformer_test.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/transformer_test.go:43]: This is intentional , it will be moved to translations in the next PR.

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/attributes.go:41]: I'd like to keep it as pointers to avoid copying entire values onto the stack for each processing step.

I am conscious that optimising early might be a bad idea without actual information of user feedback.

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/attributes.go]: Alright, apparently it isn't recommend so I have updated the actual work that it is doing.

Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/attributes.go]: > Can we add a TODO to decide what we should do if the key already exists? Should we overwrite it (as it does now) or we should we stop and return an error? This is probably more of a TODO for schema spec.

I've made it so it will prefer the schema rewrite values over the existing values to ensure it is deterministic. Meaning that values that are not part of the schema will be removed.

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (mx-psi) [inline processor/schemaprocessor/internal/migrate/attributes.go:87]: Link to an issue for deciding this in the comment?

File: processor/schemaprocessor/internal/migrate/doc.go
Comment (mx-psi) [inline processor/schemaprocessor/internal/migrate/doc.go:19]:
```suggestion
type StateSelector int
```

(and everywhere else :) )

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/attributes.go:87]: Not yet, was raised by @tigrannajaryan in a comment so I will raise it in the morning on the spec repo (or if anyone else gets to it before me).

Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/attributes.go:87]: Added here https://github.com/open-telemetry/opentelemetry-specification/issues/3497

File: processor/schemaprocessor/internal/migrate/conditional.go
Comment (tigrannajaryan) [reply processor/schemaprocessor/internal/migrate/conditional.go:35]: I see.

Instead of a string can we use `types.AttributeName`, e.g. `map[types.AttributeName]types.AttributeName`? Shouldn't this work without type casts?

I saw your [other PR ](https://github.com/open-telemetry/opentelemetry-go/pull/4082). I think I would prefer to keep the strong types in the `types` package. If it means we need to use generics here then so be it.

File: processor/schemaprocessor/internal/migrate/conditional.go
Comment (tigrannajaryan) [reply processor/schemaprocessor/internal/migrate/conditional.go]: Yes, the set to apply to (the condition) can have multiple elements. That is represented by `ca.on` field. The value to check the conditions against is only a single value (the span name or the event name). That is represented by `values` parameter, right? I don't see a situation when we can have multiple elements in `values`. 
Perhaps I am missing something, I may have misunderstood how the Apply() func is supposed to be used.

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (mx-psi) [reply processor/schemaprocessor/internal/migrate/attributes.go:87]: I think this must be solved before marking the component as alpha and adding it to our official builds, but still I don't think it should block merging this PR in the interest of moving forward with other parts of the implementation

**Review by mx-psi — APPROVED** [review]

File: processor/schemaprocessor/internal/migrate/attributes.go
Comment (tigrannajaryan) [reply processor/schemaprocessor/internal/migrate/attributes.go]: @mx-psi yes, sounds good to me.

File: processor/schemaprocessor/internal/migrate/attributes_test.go
Comment (tigrannajaryan) [reply processor/schemaprocessor/internal/migrate/attributes_test.go:110]: Can we add a TODO to fix this?

File: processor/schemaprocessor/internal/translation/revision_v1.go
Comment (tigrannajaryan) [reply processor/schemaprocessor/internal/translation/revision_v1.go:57]: Nit: since it has no particular importance I think it can be safely deleted, `values := make([]*migrate.AttributeChangeSet)` should be good enough.

**Review by tigrannajaryan — APPROVED** [review]
LGTM

File: processor/schemaprocessor/internal/translation/revision_v1.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/translation/revision_v1.go:57]: Unfortunately, (not sure when), but the latest version of go requires at least `len` and an optional `cap`.

Tested it here (to make sure it wasn't a linter getting it in the way)
https://go.dev/play/p/9uNpeSYw6gN

File: processor/schemaprocessor/internal/migrate/conditional.go
Comment (MovieStoreGuy) [reply processor/schemaprocessor/internal/migrate/conditional.go:35]: It looks like I can use `ast.AttributeMap` since that is used every where, but with the conditionals, looks like I would need the generic type for `SpanName`, `EventName`, `MetricName`.

**Comment by MovieStoreGuy** [comment]
Sorry, just removing some of the generics usage since I can use `ast.AttributeMap` where `map[AttributeKey]AttributeKey` was being defined without issue.

File: processor/schemaprocessor/internal/translation/revision_v1.go
Comment (tigrannajaryan) [reply processor/schemaprocessor/internal/translation/revision_v1.go:57]: You are right. Don't know what I was thinking. What I probably wanted to say is that you can start with an empty slice since `append` later will happily allocate as many as needed:
```go
var values []*migrate.AttributeChangeSet
```

**Comment by mx-psi** [comment]
Let's wait until EOW to give @bogdandrutu time to review and I will merge if no new concerns are raised :)

**Comment by mx-psi** [comment]
@bogdandrutu I am going to merge this, we won't move this out of 'In Development' until we clarify https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17020#discussion_r1192078334 but I am going to merge this since we have an issue tracking this on the spec. Please open follow up issues if you have any comments

**Pull request merged**
